javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+        }
+        digits(i) = ~digits(i)
+      } else {
+        for (j <- 0 until bytesLen) {
+          digits(i) = (digits(i) << 8) | (byteValues(j) & 0xFF)
+        }
+        digits(i) = -digits(i)
+      }
+    }
+  }
+
+  /**
+   * Puts a big-endian byte array into a little-endian int array.
+   */
+  private def putBytesPositiveToIntegers(byteValues: Array[Byte]) {
+    var bytesLen = byteValues.length
+    val highBytes = bytesLen & 3
+    numberLength = (bytesLen >> 2) + (if (highBytes == 0) 0 else 1)
+    digits = Array.ofDim[Int](numberLength)
+
+    // Put bytes to the int array starting from the end of the byte array
+    var i = 0
+    while (bytesLen > highBytes) {
+      digits(i) = (byteValues(bytesLen - 1) & 0xFF)       |
+                  (byteValues(bytesLen - 2) & 0xFF) << 8  |
+                  (byteValues(bytesLen - 3) & 0xFF) << 16 |
+                  (byteValues(bytesLen - 4) & 0xFF) << 24
+      bytesLen = bytesLen  -4
+      i += 1
+    }
+    // Put the first bytes in the highest element of the int array
+    for (j <- 0 until bytesLen) {
+      digits(i) = (digits(i) << 8) | (byteValues(j) & 0xFF)
+    }
+  }
+
+  /**
+   * @see BigInteger#BigInteger(String, int)
+   */
+  private def setFromString(s: String, radix: Int) {
+    var _sign: Int = 0
+    var _digits: Array[Int] = Array()
+    var _numberLength: Int = 0
+    var stringLength = s.length
+    var startChar: Int = 0
+    val endChar = stringLength
+
+    if (s.charAt(0) == '-') {
+      _sign = -1
+      startChar = 1
+      stringLength -= 1
+    } else {
+      _sign = 1
+      startChar = 0
+    }
+    /*
+     * We use the following algorithm: split a string into portions of n
+     * characters and convert each portion to an integer according to the radix.
+     * Then convert an exp(radix, n) based number to binary using the
+     * multiplication method. See D. Knuth, The Art of Computer Programming,
+     * vol. 2.
+     */
+    val charsPerInt = Conversion.digitFitInInt(radix)
+    var bigRadixDigitsLength = stringLength / charsPerInt
+    val topChars = stringLength % charsPerInt
+    if (topChars != 0) {
+      bigRadixDigitsLength += 1
+    }
+    _digits = Array.ofDim[Int](bigRadixDigitsLength)
+    val bigRadix = Conversion.bigRadices(radix - 2)
+    var digitIndex = 0
+    var substrEnd = startChar + (if (topChars == 0) charsPerInt else topChars)
+    var newDigit: Int = 0
+    var substrStart = startChar
+    while (substrStart < endChar) {
+      val bigRadixDigit = java.lang.Integer.parseInt(s.substring(substrStart, substrEnd), radix)
+      newDigit = Multiplication.multiplyByInt(_digits, digitIndex, bigRadix)
+      newDigit += Elementary.inplaceAdd(_digits, digitIndex, bigRadixDigit)
+      _digits(digitIndex) = newDigit
+      digitIndex += 1
+      substrStart = substrEnd
+      substrEnd = substrStart + charsPerInt
+    }
+    _numberLength = digitIndex
+    this.sign = _sign
+    this.numberLength = _numberLength
```

Comment:
Use `digitIndex` here directly, and remove the definition of `_numberLength` altogether.